crates/evm/traces/src/folded_stack_trace.rs

```diff
@@ -0,0 +1,301 @@
+                    step_exits.push(*step_end_idx);
+                }
+                DecodedTraceStep::Line(_) => {}
+            }
+        }
+    }
+
+    /// Exits all the previous internal calls that should end before starting step_idx.
+    fn exit_previous_steps(&mut self, step_exits: &mut Vec<usize>, step_idx: usize) {
+        let initial_length = step_exits.len();
+        step_exits.retain(|&number| number >= step_idx);
+
+        let num_exits = initial_length - step_exits.len();
+        for _ in 0..num_exits {
+            self.fst.exit();
+        }
+    }
+}
+
+/// Helps to translate a function enter-exit flow into a folded stack trace.
+///
+/// Example:
+/// fn top() { child_a(); child_b() } // consumes 500 gas
+/// fn child_a() {} // consumes 100 gas
+/// fn child_b() {} // consumes 200 gas
+///
+/// For execution of the `top` function looks like:
+/// 1. enter `top`
+/// 2. enter `child_a`
+/// 3. exit `child_a`
+/// 4. enter `child_b`
+/// 5. exit `child_b`
+/// 6. exit `top`
+///
+/// The translated folded stack trace lines look like:
+/// 1. top
+/// 2. top;child_a
+/// 3. top;child_b
+///
+/// Including the gas consumed by the function by itself.
+/// 1. top 200 // 500 - 100 - 200
+/// 2. top;child_a 100
+/// 3. top;child_b 200
+#[derive(Debug, Default)]
+pub struct FoldedStackTraceBuilder {
+    /// Trace entries.
+    traces: Vec<TraceEntry>,
+    /// Number of exits to be done before entering a new function.
+    exits: usize,
+}
+
+#[derive(Debug, Default)]
+struct TraceEntry {
+    /// Names of all functions in the call stack of this trace.
+    names: Vec<String>,
+    /// Gas consumed by this function, allowed to be negative due to refunds.
+    gas: i64,
+}
+
+impl FoldedStackTraceBuilder {
+    /// Enter execution of a function call that consumes `gas`.
+    pub fn enter(&mut self, label: String, gas: i64) {
+        let mut names = self.traces.last().map(|entry| entry.names.clone()).unwrap_or_default();
+
+        while self.exits > 0 {
+            names.pop();
+            self.exits -= 1;
+        }
+
+        names.push(label);
+        self.traces.push(TraceEntry { names, gas });
+    }
+
+    /// Exit execution of a function call.
+    pub fn exit(&mut self) {
+        self.exits += 1;
+    }
+
+    /// Returns folded stack trace.
+    pub fn build(mut self) -> Vec<String> {
+        self.subtract_children();
+        self.build_without_subtraction()
+    }
+
+    /// Internal method to build the folded stack trace without subtracting gas consumed by
+    /// the children function calls.
+    fn build_without_subtraction(&mut self) -> Vec<String> {
+        let mut lines = Vec::new();
+        for TraceEntry { names, gas } in self.traces.iter() {
+            lines.push(format!("{} {}", names.join(";"), gas));
+        }
+        lines
+    }
+
+    /// Subtracts gas consumed by the children function calls from the parent function calls.
+    fn subtract_children(&mut self) {
```

Comment:
couldn't this always handled in `enter`?

Reply:
Purposefully kept it for manual debugging in future because final FST is not very human-readable. That's why there's an intermediate check in the test cases before and after subtraction, for FST module to be self-explaining. But if it seem to complicate then I can move the logic to `enter`.